Added intial tests

anuragCES opened the pull request:
@fedor 

With respect to adding tests, I have added initial test and tapeJS config.

I need your help with ES6 based tests and testing generator. I have added "babel-tape-runner" for ES6 execution for generator testing.

Kindly review the branch.

anuragCES committed:
Added intial tests

anuragCES commented on package.json:
```diff
@@ -4,7 +4,7 @@
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "tape test.js"
```
@fedor 

Try replacing tape keyword with babel-tape-runner for using ES6 based testing.

fedor replied:
@anuragCES since there is no need for babel as I've mentioned in the comment for `.babelrc` this should be OK.

fedor commented on package.json:
```diff
@@ -33,6 +33,12 @@
   "homepage": "https://github.com/fedor/node-no-config#readme",
   "dependencies": {
-    "co": "^4.6.0"
+    "babel-core": "^6.3.15",
```
@anuragCES, for new dependencies, since they are needed only for tests, better to specify them as _development_ dependencies. To do this, use `npm install <PACKAGE> --seve-dev` (https://docs.npmjs.com/cli/install)

fedor commented on testConfig/configData.js:
```diff
@@ -0,0 +1,28 @@
+const configData = {
```
re: `testConfig/configData.js` since loading of config file is not a part of `no-config` functionality and because test data is small, I sugget to use it as object directly in `test.js` and remove `testConfig/configData.js`. Also I'm sure there would be a need for more test data (separate configs) and see no need to separate test configs from the test code and put them inside multiple files.

fedor commented on .babelrc:
```diff
@@ -0,0 +1,8 @@
+{
```
@anuragCES in current LTS version of Node.js (4.5.0) all ES6 features you use should be supported. There is no need for babel.

fedor commented on package.json:
```diff
@@ -33,6 +33,12 @@
+    "babel-preset-stage-0": "^6.3.13",
+    "babel-tape-runner": "^2.0.1",
+    "co": "^4.6.0",
+    "tap-webpack-plugin": "^1.1.0"
```
I think instead of _all_ new dependencies there should be only `tape` (`npm install tape --save`), since no need for babel and webpack support.

fedor commented on test.js:
```diff
@@ -0,0 +1,36 @@
+const co = require('co')
+const config = require('./testConfig/configData')
+const keys = ['mongo', 'redis']
+const PROCESS_ENV = process.env.NODE_ENV = 'test_env'
```
@anuragCES in the test data below I don't see `test_env` environment, only `development` and `production` are you sure it should be `test_env`?

fedor commented on test.js:
```diff
@@ -0,0 +1,36 @@
+})
+
+test('Should initialize config object', (t) => { 
+    t.plan(3)
```
@anuragCES in both test cases I suggest to use `t.end()` instead. For me it's nicer + if it's used right it will help to prevent tests from hanging if `configObject` will fail (I'll explain in the next comment)

fedor commented on test.js:
```diff
@@ -0,0 +1,36 @@
+
+test('Should initialize config object', (t) => { 
+    t.plan(3)
+    configObject.then((conf) => {
```
@anuragCES here you use `Promise` interface directly, but code would be much nicer if you'd use generators + `yield` instead. To do this use generator function wrapped by `co` for `test()` function callback:

``` js
test('Should initialize config object', co.wrap(function* (t) {
    try {
        let result = yield configObject
        // assertions
    } catch (e) {
        t.equal(e, undefined, 'No error')
    }
    t.end()
})) 
```

In the example above notice the try-catch, this helps to cover a case when `configObject` Promise got rejected, something you haven't covered. I assume in your implementation if this would happen, test code will hang (since no assertions inside `.then()` would be executed). This also shows why it's better to use `t.end()` instead of `t.plan()` (number of assertions would vary depending of `configObject` result)

fedor commented on test.js:
```diff
@@ -0,0 +1,36 @@
+    })
+})
+
+test('Should return correct values as passed input', (t) => { 
```
@anuragCES same suggestions as for the test case above

fedor commented on test.js:
```diff
@@ -0,0 +1,36 @@
+const keys = ['mongo', 'redis']
+const PROCESS_ENV = process.env.NODE_ENV = 'test_env'
+
+const configObject = co(function* () {
```
@anuragCES I don't see the point to have it. Adding 3 lines to wrap one line. Why not to use

``` js
noNodeConfig({config}).then(...
```

or

``` js
let result = yield noNodeConfig({config})
```

directly in the test cases.

fedor reviewed (changes requested):
@anuragCES,

Thanks again for pull request! Sorry for taking it so long. I realized I need time to go look thru `tape` and that I'd need to give more comments than I can put on during the work week. But finally, I had enough time to put them up.

Thanks for showing `tape`. I like its simplicity and think that I can use it by default in other projects (instead of heavy `mocha`)

Please find my comments and requests in this review.

anuragCES committed:
Completed code review changes

- Removed extra dependencies
- Removed configData as a file
- Added generators to configure the noNodeConfig object
- Added try catch to handle edge cases
- Replaced test.plan to test.end

anuragCES committed:
Removed babelrc

Not required as we’re running node version LTS 4.20

anuragCES commented:
@fedor 

Thank you for the review. I have learned a lot.

Definitely, changes have simplified the tests and assertions. I really liked the idea of passing co generator function as a parameter for tape's test.

I have made the changes, you can go ahead and review it.

Thanks for your help and support.

fedor commented on test.js:
```diff
@@ -0,0 +1,67 @@
+const test = require('tape')
```
@anuragCES, please add `'use strict'`.

Please make sure to run the code _in expected (target) environment_. I understand you've tested it with _babel_, but since babel was dropped, it makes sense to run in in _expected environment_, right? (e.g. node v4)
![image](https://cloud.githubusercontent.com/assets/2085674/18813259/acaa3974-8326-11e6-8b75-1b946f634630.png)

fedor commented on package.json:
```diff
@@ -33,6 +33,7 @@
   "dependencies": {
-    "co": "^4.6.0"
+    "co": "^4.6.0",
+    "tape": "^4.6.0"
```
@anuragCES good. One thing I've mentioned here before is a need to keep `tape` as a **dev** dependency because it's only used for its testing.

Please check this link: https://docs.npmjs.com/files/package.json#devdependencies

``` json
"dependencies": {
    ...
},
"devDependencies": {
    ...
}
```

fedor reviewed (changes requested):
@anuragCES thanks for an update, looks good! Two small changes are still needed, please find them in this review.

anuragCES committed:
Code review changes - Version 2

- Added use strict and tape as a dev dependency

- Tested with node 4.2.6

anuragCES commented:
@fedor 

Thanks for the review.

Completed changes and ready for the code review.

Run and tested it with node 4.2.6.

Let me know if you've any questions.
